source/mir/random/flex/package.d

```diff
@@ -0,0 +1,1012 @@
+                {
+                    if (fabs(hat.slope) < S(1e-10))
+                    {
+                        // reset to uniform number
+                        u /= hatArea;
+                        // pick a point on the straight line between lx and rx
+                        X = (1 - u) * lx + u * rx;
+                        goto finish;
+                    }
+                }
+                X = hat.inverse(inverseAntiderivative(u * hat.slope + antiderivative(hatLx, c), c));
+            }
+
+        finish:
+
+            immutable hatX = hat(X);
+            immutable squeezeX = squeeze(X);
+
+            auto invHatX = flexInverse(hatX, c);
+            auto invSqueezeX = squeezeArea > 0 ? flexInverse(squeezeX, c) : 0;
+
+            S u2 = uniform!("[)", S, S)(0, 1, rng);
+            immutable t = u2 * invHatX;
+
+            // u * h(c) < s(X)  "squeeze evaluation"
+            if (t <= invSqueezeX)
+                break;
+
+            // u * h(c) < f(X)  "density evaluation"
+            if (t <= pdf(X))
+                break;
+        }
+    }
+    return X;
+}
+
+/**
+Reduced version of $(LREF Interval). Contains only the necessary information
+needed in the generation phase.
+*/
+struct FlexInterval(S)
+    if (isFloatingPoint!S)
+{
+    import mir.utility.linearfun : LinearFun;
+
+    /// left position of the interval
+    S lx;
+
+    /// right position of the interval
+    S rx;
+
+    /// T_c family of the interval
+    S c;
+
+    ///
+    LinearFun!S hat;
+
+    ///
+    LinearFun!S squeeze;
+
+    ///
+    S hatArea;
+
+    ///
+    S squeezeArea;
+
+    // disallow NaN points
+    invariant {
+        import std.math : isNaN;
+        import std.meta : AliasSeq;
+        alias seq =  AliasSeq!(lx, rx, c, hatArea, squeezeArea);
+        foreach (i, v; seq)
+            assert(!v.isNaN, "variable " ~ seq[i].stringof ~ " isn't allowed to be NaN");
+
+        assert(lx < rx, "invalid interval - right side must be larger than the left side");
+    }
+}
+
+/**
+Calculate the intervals for the Flex algorithm for a T_c family given its
+density function, the first two derivatives and a valid start partitioning.
+The Flex algorithm will try to split the intervals until a chosen efficiency
+rho is reached.
+
+Params:
+    f0 = probability density function of the distribution
+    f1 = first derivative of f0
+    f1 = second derivative of f0
+    cs = T_c family (single value or array)
+    points = non-overlapping partitioning with at most one inflection point per interval
+    rho = efficiency of the Flex algorithm
+    apprMaxPoints = maximal number of splitting points before Flex is aborted
+
+Returns: Array of IntervalPoints
+*/
+FlexInterval!S[] flexIntervals(S, F0, F1, F2)
+                            (in F0 f0, in F1 f1, in F2 f2,
+                             in S[] cs, in S[] points, in S rho = 1.1,
+                             in int apprMaxPoints = 1_000, in int maxIterations = 1_000)
```

Comment:
@9il imho we should expose either `apprMaxPoints` or `maxIterations` - which one do you prefer?